### PR TITLE
Clarify the relationship of pattern type syntax to that of flex

### LINKED ISSUE #149: Pattern not support `{-}` and `{+}` operators
According to the [docs](https://docs.zeek.org/en/master/script-reference/types.html#type-pattern), Zeek accepts “patterns supported by the [flex lexical analyzer](http://westes.github.io/flex/manual/Patterns.html).”  However, character class unions/differences are not working. 

Running the following test-script with Zeek 5.0: 
```
const mypattern: pattern = /([[:print:]]{-}[[:alnum:]]foo)/;
```
Get the following error:
```
error in ././try.zeek, line 1: error compiling pattern /^?(.|\n)*(([[:print:]]{-}[[:alnum:]]foo))/
```

Reference to report on slack : https://zeekorg.slack.com/archives/CSZBXF6TH/p1662645132381609

### PR DESCRIPTION
**Pull request by ckreibich**
Let me know if this suits.

Resolves #149.

### CODE REVIEW

**Review by rsmmr — APPROVED** [review]
This probably suggests a follow up item of documenting our syntax, given that we can't just refer to flex anymore.

**Comment by ckreibich** [comment]
Yeah good point Robin. I went ahead and did that in this PR still, take a look. Hope I covered it all — I intentionally skipped the `{NAME}` feature since I don't think it's actually usable in practice and it looks like we added it in the grammar in case we use it in the future.

File: script-reference/types.rst [hunk 1]
```diff
@@ -520,15 +520,109 @@ Type Conversions
...
+    - Matches ``expr`` ``n`` or more times, where ``n`` is a non-negative integer.
+
+  * - ``<expr>{n,m}``
+    - Matches ``expr`` between ``n`` and ``m`` times, inclusively, where ``n``
+      and ``m`` are non-negative integers and ``m >= n``.
+
+  * - ``<expr1>|<expr2>``
+    - Matches either ``expr1`` or ``expr2``. In contrast to the above repetition
+      operators, expressions here span multiple characters: in ``ab+`` the ``+``
+      applies to the ``b``, while ``ab|cd`` matches ``"ab"`` or ``"cd"``.
```
Comment (rsmmr) [inline script-reference/types.rst:569]: I'd actually skip the 2nd sentence, find that more confusing than helpful.

```suggestion
    - Matches either ``expr1`` or ``expr2``. 
```

File: script-reference/types.rst [hunk 2]
```diff
@@ -520,15 +520,109 @@ Type Conversions
...
+  * - ``(<expr>)``
+    - Groups the contained expression to form a building-block of a more complex
+      one.
+
+  * - ``"<chars>"``
+    - Matches literal strings, without the quotation marks. These always match
+      as given, even if case-insensitivity is active.
+
+  * - ``[<chars>]``
+    - Defines a character class, matching any of the contained characters.
```
Comment (rsmmr) [inline script-reference/types.rst:580]: Let's add the character classes as well. Suggestion:

In addition to literal characters, the following character classes as supported as well: `[:alnum:]`,`[:alpha:]`,`[:blank:]`,`[:cntrl:]`,`[:digit:]`,`[:digit:]`,`[:graph:]`,`[:lower:]`,`[:print:]`,`[:punct:]`,`[:space:]`,`[:upper:]`,`[:xdigit:]`


They come with their standard POSIX meaning, see for example [this reference](https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions#Character_classes>).

**Review by rsmmr — APPROVED** [review]

File: script-reference/types.rst (hunk 1 above)
Comment (ckreibich) [reply script-reference/types.rst:569]: I went back and forth over that one a bunch. :-) Happy to remove it.

File: script-reference/types.rst (hunk 2 above)
Comment (ckreibich) [reply script-reference/types.rst:580]: Oh right, oversight on my part! I'll work it in.

**Pull request merged**